cloudmarker/test/test_splunkhecstore.py

```diff
@@ -0,0 +1,120 @@
+"""Tests for SplunkHECToken plugin."""
+
+import unittest
+from unittest import mock
+
+import requests
+
+from cloudmarker.stores import splunkhecstore
+
+
+class SplunkHECStoreTest(unittest.TestCase):
+    """Tests for SplunkHECStore plugin."""
+
+    def test_flush_called_once(self):
+        mock_record = {'record_type': 'firewall_rule'}
+
+        # Create a SplunkStore with buffer length 1000(default value)
+        splunk_store = splunkhecstore.SplunkHECStore('', '', '', '')
+
+        splunk_store._flush = mock.MagicMock()
+
+        splunk_store.write(mock_record)
+        splunk_store.done()
+
+        splunk_store._flush.assert_called_once_with()
+
+    def test_flush_called_twice(self):
+        mock_record = {'record_type': 'firewall_rule'}
+
+        # Create a zero length buffer
+        splunk_store = splunkhecstore.SplunkHECStore('', '', '', '', 0)
+
+        splunk_store._flush = mock.MagicMock()
+
+        splunk_store.write(mock_record)
+        splunk_store.done()
+
+        self.assertEqual(splunk_store._flush.mock_calls,
+                         [mock.call(), mock.call()])
+
+    @mock.patch('cloudmarker.stores.splunkhecstore.requests.session')
```

Comment:
It is sufficient to only `@mock.patch('requests.session')`. It is unnecessary to get access to `requests.session` by first referring to the `cloudmarker.stores.splunkhecstore` module in this case.

If we were to mock `requests` package, then we would have needed the longer form of patching, i.e., `@mock.patch('cloudmarker.stores.splunkhecstore.requests')`. Why? Because `@mock.patch('requests')` would not affect the `requests` package that has already been imported into `splunkhecstore`. In this test module, we first `import splunkhecstore` which in turn imports `requests` inside `splunkhecstore`, so `@mock.patch('requests')` in the test module does not change the `requests` package that is already imported in `splunkhecstore`. The situation looks like this:

```
Before @mock.patch('requests'):
    in test_splunkhecstore:
        requests --- refers to ---> <actual requests package object in memory>
        requests.session --- refers to ---> <actual session function in memory>
    in splunkhecstore:
        requests --- refers to ---> <the same requests package object in memory>
        requests.session --- refers to ---> <the same session function in memory>

After @mock.patch('requests'):
    in test_splunkhecstore:
        requests --- now refers to ---> (a MagicMock object in memory)
        requests.session --- now refers to ---> (a dynamically created MagicMock object in memory)
    in splunkhecstore:
        requests --- still refers to ---> <actual requests package object in memory>
        requests.session --- still refers to ---> <actual session function in memory>
```

Note that in the beginning, both `test_splunkhecstore` and `splunkhecstore` refer to the same `requests` package object imported in memory. That's because a package/module is imported only once in the lifetime of a project regardless of how many times we have imported them in different parts of the project with the `import` statement.

Then note that after mocking, the `splunkhecstore`'s `requests` reference is still referring to the original `requests` even after mocking.

But in this test module since, we are mocking `requests.session` here, not `requests`, the situation turns out to be different:

```
Before @mock.patch('requests.session'):
    in test_splunkhecstore:
        requests --- refers to ---> <actual requests package object in memory>
        requests.session --- refers to ---> <actual session function in memory>
    in splunkhecstore:
        requests --- refers to ---> <the same requests package object in memory>
        requests.session --- refers to ---> <the same session function in memory>

After @mock.patch('requests.session'):
    test_splunkhecstore:
        requests --- still refers to ---> <actual requests package object in memory>
        requests.session --- now refers to ---> (a MagicMock object in memory)
    in splunkhecstore:
        requests --- still refers to ---> <actual requests package object in memory>
        requests.session --- now refers to ---> (a MagicMock object in memory)
```

With `mock.patch('requests.session')`, in the test module, we are referring to the same `requests` package object that `splunkhecstore` is also referring to, so when we replace its `session` function, i.e., `requests.session` with a mock, `splunkhecstore` also gets the same mock when it tries to use `requests.session`.

Summary: We can use just `@mock.patch('requests.session')` in this test module.

Reply:
👍